File: java/data_structures/priority_queue/custom/PriorityQueueCustom.java
package priority_queue.custom;

import heap.generic.*;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Priority Queue implementation using the existing MaxHeap.
 *
 * <p>Each element is stored as a PriorityItem (priority, -order, value) to ensure:
 * <ul>
 *   <li>Higher priority values are served first (max-priority behavior)</li>
 *   <li>FIFO stability between elements with the same priority (via negative order)</li>
 * </ul>
 *
 * <p>Time Complexity:
 * <ul>
 *   <li>push(): O(log n)</li>
 *   <li>pop():  O(log n)</li>
 *   <li>peek(): O(1)</li>
 *   <li>isEmpty(), size(): O(1)</li>
 * </ul>
 */
public class PriorityQueueCustom<T> {
    private final MaxHeap<PriorityItem<T>> heap;
    private final AtomicLong counter; 

    public PriorityQueueCustom() {
        this.heap = new MaxHeap<>();
        this.counter = new AtomicLong(0);
    }

    public void push(T value, int priority) {
        long order = counter.getAndIncrement();
        // Negative order ensures FIFO among elements with equal priority
        heap.insert(new PriorityItem<>(priority, -order, value));
    }

    public T pop() {
        if (isEmpty()) {
            throw new IllegalStateException("pop from an empty priority queue");
        }
        PriorityItem<T> item = heap.pop();
        return item.getValue();
    }

    public T peek() {
        if (isEmpty()) {
            throw new IllegalStateException("peek from an empty priority queue");
        }
        return heap.peek().getValue();
    }

    public boolean isEmpty() {
        return size() == 0;
    }

    public int size() {
        return heap.size();
    }

    private static class PriorityItem<T> implements Comparable<PriorityItem<T>> {
        private final int priority;
        private final long negOrder;
        private final T value;

        public PriorityItem(int priority, long negOrder, T value) {
            this.priority = priority;
            this.negOrder = negOrder;
            this.value = value;
        }

        public T getValue() {
            return value;
        }

        @Override
        public int compareTo(PriorityItem<T> other) {
            // Higher priority first
            if (this.priority != other.priority) {
                return Integer.compare(this.priority, other.priority);
            }
            // For equal priority, the higher negOrder (smaller insertion order) comes first
            return Long.compare(this.negOrder, other.negOrder);
        }

        @Override
        public String toString() {
            return String.format("(%d, %d, %s)", priority, negOrder, value);
        }
    }
}

File: csharp/data_structure/priority_queue/custom/PriorityQueueCustom.cs
using System;
using System.Threading;
using DataStructure.Generic.Heap;

namespace DataStructure.PriorityQueue.Custom;

/// <summary>
/// Priority Queue implementation using a custom MaxHeap.
/// 
/// Each element is stored as a <see cref="PriorityItem{T}"/> (priority, -order, value) to ensure:
/// - Higher priority values are served first (max-priority behavior)
/// - FIFO stability among elements with the same priority (via negative order)
/// 
/// Time Complexity:
/// - Push(): O(log n)
/// - Pop():  O(log n)
/// - Peek(): O(1)
/// - IsEmpty(), Size(): O(1)
/// </summary>
public class PriorityQueueCustom<T>
{
    private readonly MaxHeap<PriorityItem<T>> _heap;
    private long _counter; 

    public PriorityQueueCustom()
    {
        _heap = new MaxHeap<PriorityItem<T>>();
        _counter = 0;
    }


    public void Push(T value, int priority)
    {
        long order = Interlocked.Increment(ref _counter);
        _heap.Insert(new PriorityItem<T>(priority, -order, value));
    }

    public T Pop()
    {
        if (IsEmpty())
            throw new InvalidOperationException("Pop from an empty priority queue");

        var item = _heap.Pop();
        return item.Value;
    }

    public T Peek()
    {
        if (IsEmpty())
            throw new InvalidOperationException("Peek from an empty priority queue");

        return _heap.Peek().Value;
    }

    public bool IsEmpty() => Size() == 0;

    public int Size() => _heap.Size();

    private sealed class PriorityItem<TValue> : IComparable<PriorityItem<TValue>>
    {
        public int Priority { get; }
        public long NegOrder { get; }
        public TValue Value { get; }

        public PriorityItem(int priority, long negOrder, TValue value)
        {
            Priority = priority;
            NegOrder = negOrder;
            Value = value;
        }

        public int CompareTo(PriorityItem<TValue> other)
        {
            if (other == null) return 1;

            int cmp = Priority.CompareTo(other.Priority);
            if (cmp != 0)
                return cmp;
            return NegOrder.CompareTo(other.NegOrder);
        }

        public override string ToString() => $"({Priority}, {NegOrder}, {Value})";
    }
}
